### PR TITLE
Das 2192

### PR DESCRIPTION
**Pull request by sudha-murthy**
## Description

Prefetch of dimension variables was occurring for variable subsetting for SMAP L3/L4 collections. The update to the is_index_subset method fixes the bug.

## Jira Issue ID
DAS-2192

## Local Test Steps
- Build hoss image locally after fetching the branch DAS-2192
- All the following 8 SMAP L3 requests should succeed and the logs should not have a call for 'prefetch'
1) SPL3SMP.009
http://localhost:3000/C1268452378-EEDTEST/ogc-api-coverages/1.0.0/collections/parameter_vars/coverage/rangeset?forceAsync=true&granuleId=G1268452388-EEDTEST&variable=Soil_Moisture_Retrieval_Data_AM%2Falbedo&skipPreview=true
2) SPL3SMP_E.006
http://localhost:3000/C1268399578-EEDTEST/ogc-api-coverages/1.0.0/collections/parameter_vars/coverage/rangeset?forceAsync=true&granuleId=G1268399648-EEDTEST&variable=Soil_Moisture_Retrieval_Data_AM%2Falbedo&skipPreview=true
3) SPL3FTP.004
 http://localhost:3000/C1268617120-EEDTEST/ogc-api-coverages/1.0.0/collections/parameter_vars/coverage/rangeset?forceAsync=true&granuleId=G1268617167-EEDTEST&variable=Freeze_Thaw_Retrieval_Data_Global%2Faltitude_dem&skipPreview=true
4) SPL3FTP_E.004
 http://localhost:3000/C1268616149-EEDTEST/ogc-api-coverages/1.0.0/collections/parameter_vars/coverage/rangeset?forceAsync=true&granuleId=G1268616387-EEDTEST&variable=Freeze_Thaw_Retrieval_Data_Global%2Faltitude_dem&skipPreview=true
5) SPL3FTA.003
 http://localhost:3000/C1268612485-EEDTEST/ogc-api-coverages/1.0.0/collections/all/coverage/rangeset?forceAsync=true&granuleId=G1268612495-EEDTEST&skipPreview=true
6) SPL3SMAP.003
 http://localhost:3000/C1268612473-EEDTEST/ogc-api-coverages/1.0.0/collections/parameter_vars/coverage/rangeset?forceAsync=true&granuleId=G1268612483-EEDTEST&variable=Soil_Moisture_Retrieval_Data%2Falbedo&skipPreview=true
7) SPL2SMAP_S.003
 http://localhost:3000/C1268429762-EEDTEST/ogc-api-coverages/1.0.0/collections/parameter_vars/coverage/rangeset?forceAsync=true&granuleId=G1268612499-EEDTEST&variable=a%2FLogFileName%2CSoil_Moisture_Retrieval_Data_1km%2Falbedo_1km&skipPreview=true
8) SPL3SMA.003
 http://localhost:3000/C1268429762-EEDTEST/ogc-api-coverages/1.0.0/collections/parameter_vars/coverage/rangeset?forceAsync=true&granuleId=G1268612499-EEDTEST&variable=a%2FLogFileName%2CSoil_Moisture_Retrieval_Data_1km%2Falbedo_1km&skipPreview=true

## PR Acceptance Checklist
* [ ] Jira ticket acceptance criteria met.
* [X ] `CHANGELOG.md` updated to include high level summary of PR changes.
* [X] `docker/service_version.txt` updated if publishing a release.
* [X] Tests added/updated and passing.
* [ ] Documentation updated (if needed).

The fix does only corrects variable subsetting. Spatial subsetting will be fixed after a fix to varinfo in DAS-2231
The jupyter notebook tests will be updated after the fix to DAS-2231 and the corresponding HOSS changes

### CODE REVIEW

File: hoss/dimension_utilities.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ def is_index_subset(message: Message) -> bool:
 
     """
     return any(
-        rgetattr(message, subset_parameter, None) is not None
+        rgetattr(message, subset_parameter, None) not in (None, [])
```
Comment (owenlittlejohns) [inline hoss/dimension_utilities.py:48]: I think we need a new unit test (or more likely a sub test) here to make sure the empty list is correctly identified in this check. Probably also a sub test with a non-empty list, too, just to be sure.

[Here's the test for this function](https://github.com/nasa/harmony-opendap-subsetter/blob/main/tests/unit/test_dimension_utilities.py#L623). I think the easiest thing is to just add two new subtests.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,11 @@
+
+## v1.0.5
+### 2024-08-19
+
+This version of HOSS updates the 'is_index_subset' method to check for empty list (in case of dimensions subset)
```
Comment (owenlittlejohns) [inline CHANGELOG.md:5]: Super pedantic nit: The single-quotes around the `is_index_subset` should probably be backticks so that the function renders as code in the markdown.

**Review by owenlittlejohns — CHANGES_REQUESTED** [review]
The change makes sense. The two main things are adding a couple of quick subtests to make sure the functionality works as expected and making sure the [pre-commit checks](https://results.pre-commit.ci/run/github/648685054/1724104811.Nd21CLS0Qzu3kBCyZDI1TA) are passing. There are instructions [here](https://github.com/nasa/harmony-opendap-subsetter/blob/main/README.md#pre-commit-hooks) to set up `pre-commit` for the repository.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,11 @@
+
+## v1.0.5
+### 2024-08-19
+
+This version of HOSS updates the 'is_index_subset' method to check for empty list (in case of dimensions subset) 
+as well as None (for the spatial, bbox and temporal subsets. This prevents prefetch from being called for
```
Comment (owenlittlejohns) [inline CHANGELOG.md:6]: Other nit: This line is missing a closing parenthesis. I think it's meant to be:

> as well as `None` (for spatial, bounding box and temporal subsets). This prevents prefetch from being called for

I guess one other note: I don't know that "prefetch" is too informative for a lot of folks. Maybe:

> This prevents 1-D dimension variables from being unnecessarily requested from OPeNDAP for some variable subsets, in particular SMAP L3 and L4 collections.

Nit on line below: "occuring" -> "occurring".

File: hoss/dimension_utilities.py (hunk 1 above)
Comment (sudha-murthy) [reply hoss/dimension_utilities.py:48]: I think there are several subtests with non empty list. But I added 2 subtests for the empty and non empty list cases

File: tests/unit/test_dimension_utilities.py [hunk 4]
```diff
@@ -679,6 +679,38 @@ def test_is_index_subset(self):
                 )
             )
 
+        with self.subTest('No index Subset'):
```
Comment (owenlittlejohns) [inline tests/unit/test_dimension_utilities.py:682]: Nitpick: I think it's probably worth explicitly stating this subtest is for an empty `dimensions` rather than "No index subset".

File: hoss/dimension_utilities.py (hunk 1 above)
Comment (owenlittlejohns) [reply hoss/dimension_utilities.py:48]: Yeah, looking in more detail, I think you are right about the non-empty list. But thanks for adding a couple of unit tests.

**Review by owenlittlejohns — COMMENTED** [review]
Thanks for adding the unit test. That looks good.

I think all I have left are the nits from `CHANGELOG.md` (missing closing parenthesis and adding a bit more context to the prose about the prefetch request to OPeNDAP). Other than that, I found one more nit about the description of the new unit test. Otherwise, this is looking good.

For what it's worth, the reason I'm being a bit pedantic about the `CHANGELOG.md` is it forms the basis of the public release notes for 1.0.5 release, which are are pretty public-facing, so it's good to make sure there's not typos in them etc.

**Review by autydp — APPROVED** [review]
This change + the unit test updates look good and cover the necessary updates.  Good work!

**Comment by sudha-murthy** [comment]
> This change + the unit test updates look good and cover the necessary updates. Good work!

Thanks @autydp

**Comment by sudha-murthy** [comment]
> Thanks for adding the unit test. That looks good.
> 
> I think all I have left are the nits from `CHANGELOG.md` (missing closing parenthesis and adding a bit more context to the prose about the prefetch request to OPeNDAP). Other than that, I found one more nit about the description of the new unit test. Otherwise, this is looking good.
> 
> For what it's worth, the reason I'm being a bit pedantic about the `CHANGELOG.md` is it forms the basis of the public release notes for 1.0.5 release, which are are pretty public-facing, so it's good to make sure there's not typos in them etc.

Thanks @owenlittlejohns 
I think I updated them both now. I must have missed committing the CHANGELOG.md before.

**Review by owenlittlejohns — APPROVED** [review]
Thanks for addressing the comments I had.

When merging please use the "Squash and Merge" option, and be sure to prepend the squashed commit message with the ticket number, so something like "DAS-2192 - Ensure HOSS correctly identifies when an index range subset is necessary". (The commits so far in this PR do not do this)

**Review by flamingbear — APPROVED** [review]
Looks good, I was following along in the background.

**Comment by sudha-murthy** [comment]
"DAS-2192 - Ensure HOSS correctly identifies when an index range subset is necessary".
